sql/core/src/test/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInSortMergeJoinSuite.scala

```diff
@@ -0,0 +1,141 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.bucketing
+
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.plans.Inner
+import org.apache.spark.sql.execution.{BinaryExecNode, FileSourceScanExec, SparkPlan}
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, InMemoryFileIndex}
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.execution.joins.{BroadcastHashJoinExec, BuildLeft, SortMergeJoinExec}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.{SharedSparkSession, SQLTestUtils}
+import org.apache.spark.sql.types.{IntegerType, StructField, StructType}
+
+class CoalesceBucketsInSortMergeJoinSuite extends SQLTestUtils with SharedSparkSession {
+  private def newRelation(numBuckets: Int): HadoopFsRelation = HadoopFsRelation(
+    location = new InMemoryFileIndex(spark, Nil, Map.empty, None),
+    partitionSchema = StructType(Seq(StructField("a", IntegerType))),
+    dataSchema = StructType(Seq(StructField("a", IntegerType))),
+    bucketSpec = Some(BucketSpec(numBuckets, Seq("a"), Seq("a"))),
+    fileFormat = new ParquetFileFormat(),
+    options = Map.empty)(spark)
+
+  private def run(
+      leftNumBuckets: Int,
+      rightNumBuckets: Int,
+      isSortMergeJoin: Boolean,
+      expectedLeftCoalescedNumBuckets: Option[Int],
+      expectedRightCoalescedNumBuckets: Option[Int]): Unit = {
+    val leftRelation = newRelation(leftNumBuckets)
+    val rightRelation = newRelation(rightNumBuckets)
+    val left = FileSourceScanExec(
+      leftRelation, Nil, leftRelation.dataSchema, Nil, None, None, Nil, None)
+    val right = FileSourceScanExec(
+      rightRelation, Nil, rightRelation.dataSchema, Nil, None, None, Nil, None)
+    val join = if (isSortMergeJoin) {
+      SortMergeJoinExec(Nil, Nil, Inner, None, left, right)
+    } else {
+      BroadcastHashJoinExec(Nil, Nil, Inner, BuildLeft, None, left, right)
+    }
+
+    val plan = CoalesceBucketsInSortMergeJoin(spark.sessionState.conf)(join)
+
+    def verify(expected: Option[Int], subPlan: SparkPlan): Unit = {
+      val coalesced = subPlan.collect {
+        case f: FileSourceScanExec if f.optionalNumCoalescedBuckets.nonEmpty =>
+          f.optionalNumCoalescedBuckets.get
+      }
+      if (expected.isDefined) {
+        assert(coalesced.size == 1 && coalesced(0) == expected.get)
+      } else {
+        assert(coalesced.isEmpty)
+      }
+    }
+
+    verify(expectedLeftCoalescedNumBuckets, plan.asInstanceOf[BinaryExecNode].left)
+    verify(expectedRightCoalescedNumBuckets, plan.asInstanceOf[BinaryExecNode].right)
+  }
+
+  test("bucket coalescing - basic") {
+    withSQLConf(SQLConf.COALESCE_BUCKETS_IN_JOIN_ENABLED.key -> "true") {
+      run(
+        leftNumBuckets = 4,
+        rightNumBuckets = 8,
```

Comment:
Could you run the opposite condition (left=8, right=4 in this case) in all the tests for test coverage (https://github.com/apache/spark/pull/28123/files#diff-771ed8520558b80b80cfd4f222b4dc21R99-R103)?